File: practice/leetcode/python/128_longest_consecutive_sequence.py
#https://leetcode.com/problems/longest-consecutive-sequence/description/
#Given an unsorted array of integers nums, return the length of 
#the longest consecutive elements sequence.
#You must write an algorithm that runs in O(n) time.
from typing import List


class Solution:
    def longestConsecutuveSequence(self, nums):
        nums = set(nums)
        best = 0

        for x in nums:
            while x - 1 not in nums:
                y = x + 1
                while y in nums:
                    y += 1
                best = max(best, y - x)
        return best

    def longestConsecutiveSequence(self, nums: List[int]) -> int:
        numsSet = set(nums)
        max_length = 0

        #for every num find whether its the smallest number if not then ignore it
        for num in nums:
            if (num - 1) in numsSet:
                continue

            #now you have found new smallest number now find its all consecutive numbers present in the list
            y = num + 1
            while y in nums:
                y += 1

            max_length = max(max_length, y - num)

        return max_length

File: practice/leetcode/go/128_longest_consecutive_sequence.go
//https://leetcode.com/problems/longest-consecutive-sequence/description/
//Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
//You must write an algorithm that runs in O(n) time.
func longestConsecutiveSequence(nums []int) int {
	set := make(map[int]bool, 0)
	result := 0
	for _, num := range nums {
		set[num] = true
	}
	for num := range set {
		if set[num - 1] {
			continue
		}
		count := 1
		for i = num + 1; set[i]; i++ {
			count++
		}
		result = max(result, count)
	}
	return result
}
// longestConsecutiveSequence2
func longestConsecutiveSequence2(nums []int) int {
	set := make(map[int]bool, 0)
	result := 0

	for _, num := range nums {
		if set[num] = true
	}
	for num := range set {
		if set[num-1] {
			continue
		}
		count := 1
		for i = num + 1; set[i]; i++ {
			count++
		}
		result = max(result, count)
	}
	return(result)
}
